### PR TITLE
creates example/compiler/01_iet.ipynb

### PR DESCRIPTION
**Pull request by btrb**

### CODE REVIEW

**Comment by FabioLuporini** [comment]
OK, perhaps this is even too small :) you can put some initial content and then we can provide some feedback

**Comment by beterraba** [comment]
Sure, sure. I was just sanity-checking if Travis likes my repo. I didn't know he takes that long by the way :0

**Comment by FabioLuporini** [comment]
oh yes, it's because of our continuous integration (CI) systems, which run the entire test suite on multiple platforms to verify that none of the changes have broken anything. This needs some times

**Comment by beterraba** [comment]
Hey Fabio, thank you for your comments.
I'm not sure if we're seeing the same version of the updates.
In my current version I don't use the world `scalar` at any point.
I also state about such a particularity of IETs. 

About Travis' tests, here I read:  
`All checks have passed
2 successful checks`

Let me try to push that again.

**Comment by beterraba** [comment]
Please let me know if that worked. 

Just for sanity checking, please let me share the steps I took after editing the file `01_iet.ipynb`:

`git checkout master`
`git pull`
`git checkout iet-docummentation`
`git merge master`
`git push origin iet-docummentation`

File: examples/compiler/01_iet.ipynb [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This tutorial describes one particular intermediate representation used by the Devito Compiler: the Iteration/Expression Tree (IET), a special type of Abstract Syntax Tree (AST).\n",
+    "\n",
+    "# Part I - Top Down\n",
+    "\n",
+    "First, let's describe a $Domain$, and a $Function$ that will allow us to specify how such a domain gets modified. In particular, we will look at Functions that change through $time$. \n",
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: I suggest to use:

- \` ... \` for Devito objects, such as Function and Operator, so if you write \`Function\`  it gets rendered as `Function`
- \_ ... \_ for relevant names, such as "domain". This gets rendered to italics

Use dollars only for mathematical symbols, which should never be the case in these tutorials.

I'm writing this here, but it should be applied to both notebooks

File: examples/compiler/01_iet.ipynb [hunk 2]
```diff
@@ -0,0 +1,696 @@
...
+    "\n",
+    "Thus we need a Function object with which we can build a timestepping scheme. For this purpose Devito provides so-called `TimeData` objects that encapsulate Functions that are differentiable in space and time, which are derived from basic $SymPy$ Functions. \n",
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: \`Function\`

File: examples/compiler/01_iet.ipynb [hunk 3]
```diff
@@ -0,0 +1,696 @@
...
+    "\n",
+    "With this we can derive symbolic expressions for the backward derivatives in space directly via the `u.dxl` and `u.dyl` shorthand expressions (the l indicates \"left\" or backward differences) and the shorthand notation `u.dt` provided by TimeData objects to derive the forward derivative in time.\n",
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: "TimeData" is obsolete syntax. Perhaps from more than one year ago. Use \`TimeFunction\` instead. Also, check if it appears somewhere else...

File: examples/compiler/01_iet.ipynb [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+    "grid = Grid(shape=(3, 3))\n",
+    "u = TimeFunction(name='u', grid=grid)\n",
+    "u.data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here, we have just declared a two-dimensional Domain with three coordinates at each dimension ($x$ and $y$). Each coordinate of such a discrete space will be holding a $real~value$.\n",
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: domain

Comment (FabioLuporini) [reply examples/compiler/01_iet.ipynb]: no need for emphasis on real-value

File: examples/compiler/01_iet.ipynb [hunk 5]
```diff
@@ -0,0 +1,696 @@
...
+    "\n",
+    "As we can see, we can always access the values at each coordinate of the Domain. At this point, no modifications have been done to it yet. `u.data` give us a quick access over the values holded by each cell of such a Domain. `u.data[0]` holds the values in the grid at the \"current\" iteration time, given a time-step to be considered, whereas `u.data[1]` holds the values of `u` for the \"current+1\" time-step.\n",
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: domain, not Domain

File: examples/compiler/01_iet.ipynb [hunk 6]
```diff
@@ -0,0 +1,696 @@
...
+    "\n",
+    "We can now create an `Operator` that will perform modifications onto our Domain according to $differential~ equations$ through a computational stencil. \n",
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: no need for emphasis on differential-equations , and no need for the hyphen

File: examples/compiler/01_iet.ipynb [hunk 7]
```diff
@@ -0,0 +1,696 @@
...
+    "It means that those differential equations will be translated into finite differences that will be used to update the values at each coordinate of the space.\n",
+    "Such finite differences, or `expressions`, will be applied for specific ranges of `iterations` over the Domain. "
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: no need for \` wrapping "iterations", unless "iterations" is a variable name that you've defined in one of the cells (I haven't checked)

also, Domain -> domain

File: examples/compiler/01_iet.ipynb [hunk 8]
```diff
@@ -0,0 +1,696 @@
...
+    "eq = Eq(u.forward, u+1)\n",
+    "op = Operator(eq)\n",
+    "op.args['expressions']"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For instance, the particular `Equation` object above allows us say that, at each time step, `1` will be added to every position of the domain.\n",
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: Eq

Comment (FabioLuporini) [reply examples/compiler/01_iet.ipynb]: no need for \` wrapping 1

File: examples/compiler/01_iet.ipynb [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+    "\n",
+    "Let's take a look at the $kernel$ that will be used to compute how this equation alters the domain."
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: dollars

File: examples/compiler/01_iet.ipynb [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+    "u2.data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note, however, that there is no need for recompilation. Just-in-time (JIT) compilation occurs only once, triggered by the first execution.\n",
+    "\n",
+    "In summary, an `Operator` takes as input an ordered sequence of `SymPy` equations and performs $indexification$, $substitution$ and $domain-alignment$ for building $lowered~equations$. The lowered equations are analyzed to collect information relevant for the `Operator` construction and execution, resulting in `Clusters` of ordered sequences of equations having the same $iteration~space$ (`ISpace`) and no dimension-carried anti-dependences among them. Those `Clusters` go through $symbolic-optimization$ by the Devito Symbolic Engine (DSE), which consists of a series of passes, ranging from standard commmon sub-expression elimination (CSE) to more advanced rewrite procedures applied to each `Cluster`. The result is a new ordered sequence of `Clusters` that are lowered to an Iteration/Expression Tree (IET).\n",
```
Comment (FabioLuporini) [inline examples/compiler/01_iet.ipynb]: no need for dollars anywhere

**Comment by FabioLuporini** [comment]
The cell execution order is messed up. At some point a ``print(op)`` failed. We need to add these notebooks to CI so that breakages are captured. IOW, we have to add some lines to ``.travis.yml`` and ``Jenkinsfile``.

File: examples/compiler/01_iet.ipynb [hunk 11]
```diff
@@ -0,0 +1,664 @@
...
+   "source": [
+    "This tutorial describes one particular intermediate representation used by the Devito Compiler: the Iteration/Expression Tree (IET), a special type of Abstract Syntax Tree (AST).\n",
+    "\n",
+    "# Part I - Top Down\n",
+    "\n",
+    "First, let's describe a _domain_, and a _function_ that will allow us to specify how such a domain gets modified. In particular, we will look at Functions that change through _time_. \n",
+    "\n",
+    "Thus we need a `Function` object with which we can build a timestepping scheme. For this purpose Devito provides so-called `TimeFunction` objects that encapsulate Functions that are differentiable in space and time, which are derived from basic _SymPy_ Functions. \n",
+    "\n",
+    "With this we can derive symbolic expressions for the backward derivatives in space directly via the `u.dxl` and `u.dyl` shorthand expressions (the l indicates \"left\" or backward differences) and the shorthand notation `u.dt` provided by `TimeFunction` objects to derive the forward derivative in time.\n",
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: `u.dl` is not backward derivative, it is a derivative centered left of the current point, you can see it for high orders and you'll see it is not. `backward/forward` or `upwind/downwind` derivatives don't have shortcuts yet, namings needs some fix too.

File: examples/compiler/01_iet.ipynb [hunk 12]
```diff
@@ -0,0 +1,664 @@
...
+    "grid = Grid(shape=(3, 3))\n",
+    "u = TimeFunction(name='u', grid=grid)\n",
+    "u.data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here, we have just declared a two-dimensional domain with three coordinates at each dimension (_x_ and _y_). Each coordinate of such a discrete space will be holding a real value.\n",
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: Not sure what this mean. 
> "two-dimensional domain with three coordinates" 
doesn't make sense
> Each coordinate of such a discrete space will be holding a real value.
A coordinate does not hold values, a Function holds values at given coordinates.

File: examples/compiler/01_iet.ipynb [hunk 13]
```diff
@@ -0,0 +1,664 @@
...
+    "\n",
+    "As we can see, we can always access the values at each coordinate of the domain. At this point, no modifications have been done to it yet. `u.data` give us a quick access over the values holded by each cell of such a domain. `u.data[0]` holds the values in the grid at the \"current\" iteration time, given a time-step to be considered, whereas `u.data[1]` holds the values of `u` for the \"current+1\" time-step.\n",
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: - values of the Function.
- held by each node, not cell (typo too)
- `u.data[0]` would put `u.data[0, :,:]` for consistency with coordinates discussion above.
- "given a time-step to be considered" => remove intricated

File: examples/compiler/01_iet.ipynb [hunk 14]
```diff
@@ -0,0 +1,664 @@
...
+    "\n",
+    "We can now create an `Operator` that will perform modifications onto our domain according to differential equations through a computational stencil. \n",
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: modification to the Function, not the domain, the domain never changes

File: examples/compiler/01_iet.ipynb [hunk 15]
```diff
@@ -0,0 +1,664 @@
...
+    "It means that those differential equations will be translated into finite differences that will be used to update the values at each coordinate of the space.\n",
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: at each spatial and temporal coordinates

File: examples/compiler/01_iet.ipynb [hunk 16]
```diff
@@ -0,0 +1,664 @@
...
+    "eq = Eq(u.forward, u+1)\n",
+    "op = Operator(eq)\n",
+    "op.args['expressions']"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For instance, the particular `Eq` object above allows us say that, at each time step, 1 will be added to every position of the domain.\n",
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: This is not what this stencil do. This stencil updates the current time step with the value at the previous time-step + 1.
> at each time step, 1 will be added to every position of the domain.

this would be `Eq(u, u + 1)`

File: examples/compiler/01_iet.ipynb [hunk 17]
```diff
@@ -0,0 +1,664 @@
...
+    "\n",
+    "Let's take a look at the _kernel_ that will be used to compute how this equation alters the domain."
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: kernel not defined. What's a kernel

File: examples/compiler/01_iet.ipynb [hunk 18]
```diff
@@ -0,0 +1,664 @@
...
+   ],
+   "source": [
+    "print(op)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now that we have an `Operator` set up, we are ready to update our domain throught the apply method. Without additional parameters specified, the `Operator` runs on the same data objects used to build it. It is important to stress that, regarded that this operator is buffered by a `TimeFunction`, the maximum iteration point along the time dimension must be explicitly specified (otherwise, the `Operator` wouldn't know how many iterations to run).\n",
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: buffered not defined again, too vague by itself.

File: examples/compiler/01_iet.ipynb [hunk 19]
```diff
@@ -0,0 +1,664 @@
...
+    "\n",
+    "Notice that no modifications to the domain have been done so far. To verify that, query for `u.data`."
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: Function not domain, to change throughout

File: examples/compiler/01_iet.ipynb [hunk 20]
```diff
@@ -0,0 +1,664 @@
...
+   "source": [
+    "op.apply(time=2)\n",
+    "u.data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This is the first time that we are invoking the method `apply` from `Operator`. Therefore, the kernel that we saw will get written in a `.c` file, and compiled into a `.so` lib if `DEVITO_BACKEND` is set to `core`.\n",
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: into a library with corresponding extension name (.so, .dylib,...)

File: examples/compiler/01_iet.ipynb [hunk 21]
```diff
@@ -0,0 +1,664 @@
...
+    "\n",
+    "Then, as no key-value parameters are specified, the `Operator` runs with its default arguments, namely `u=u, x_m=0, x_M=2, y_m=0, y_M=2,` and `time_m=0`. The subindexes `m` and `M` stands for the minimum and the maximum at those specific Dimensions, respectivelly. Thus `time_M` will be set to `2`, here.\n",
+    "\n",
+    "At this point, the same `Operator` can be used for a completely different run. Let's create another `TimeFunction` for governing a similar domain. "
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: doesn't make sense, rephrase.

File: examples/compiler/01_iet.ipynb [hunk 22]
```diff
@@ -0,0 +1,664 @@
...
+    "u2.data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note, however, that there is no need for recompilation. Just-in-time (JIT) compilation occurs only once, triggered by the first execution.\n",
+    "\n",
+    "In summary, an `Operator` takes as input an ordered sequence of `SymPy` equations and performs _indexification_, _substitution_ and _domain-alignment_ for building _lowered equations_. The lowered equations are analyzed to collect information relevant for the `Operator` construction and execution, resulting in `Clusters` of ordered sequences of equations having the same _iteration space_ (`ISpace`) and no dimension-carried anti-dependences among them. Those `Clusters` go through _symbolic-optimization_ by the Devito Symbolic Engine (DSE), which consists of a series of passes, ranging from standard commmon sub-expression elimination (CSE) to more advanced rewrite procedures applied to each `Cluster`. The result is a new ordered sequence of `Clusters` that are lowered to an Iteration/Expression Tree (IET).\n",
```
Comment (mloubout) [inline examples/compiler/01_iet.ipynb]: - `Devito` equation

>_indexification_, _substitution_ and _domain-alignment_ for building _lowered equations_.

details

Very dense. WOuld look at @fablup paper or at list put ref. This is hard to follow by itself.

**Comment by beterraba** [comment]
Is there a way to test if they pass into Travis and Jenkins locally? Here I see the green light that "All check have passed".
I have recently pulled the new master and merged it into this branch. Should I do it again and re-push this commit?

**Comment by beterraba** [comment]
Never mind the above question. I think I got what you meant.

File: .travis.yml [hunk 23]
```diff
@@ -50,6 +50,16 @@ matrix:
...
+    - os: linux
+      python: "3.6"
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test     # For gcc 4.9, 5 and 7
+          packages:
+            - gcc-7
+            - g++-7
+      env: DEVITO_ARCH=gcc-7 DEVITO_OPENMP=0 DEVITO_BACKEND=ops INSTALL_TYPE=conda RUN_EXAMPLES=False
```
Comment (mloubout) [inline .travis.yml]: Why is this here?

Comment (beterraba) [reply .travis.yml]: Sorry, brain damage.

**Comment by FabioLuporini** [comment]
Closing as moved to #667
